auth/casbin_store.go

```diff
@@ -0,0 +1,224 @@
+	tx := as.s.be.BatchTx()
+	tx.Lock()
+	defer tx.Unlock()
+
+	role := getRole(tx, r.Role)
+	if role == nil {
+		return nil, ErrRoleNotFound
+	}
+
+	if !as.enforcer.HasPermissionForUser(r.Role, r.Key, r.RangeEnd) {
+		return nil, ErrPermissionNotGranted
+	}
+
+	as.enforcer.DeletePermissionForUser(r.Role, r.Key, r.RangeEnd)
+
+	plog.Noticef("revoked key %s from role %s", r.Key, r.Role)
+	return &pb.AuthRoleRevokePermissionResponse{}, nil
+}
+
+func (as *casbinAuthStore) RoleGrantPermission(r *pb.AuthRoleGrantPermissionRequest) (*pb.AuthRoleGrantPermissionResponse, error) {
+	tx := as.s.be.BatchTx()
+	tx.Lock()
+	defer tx.Unlock()
+
+	role := getRole(tx, r.Name)
+	if role == nil {
+		return nil, ErrRoleNotFound
+	}
+
+	if as.enforcer.HasPermissionForUser(r.Name, string(r.Perm.Key), string(r.Perm.RangeEnd), authpb.Permission_Type_name[int32(r.Perm.PermType)]) {
+		return nil, ErrPermissionAlreadyExist
+	}
+
+	as.enforcer.AddPermissionForUser(r.Name, string(r.Perm.Key), string(r.Perm.RangeEnd), authpb.Permission_Type_name[int32(r.Perm.PermType)])
+
+	plog.Noticef("role %s's permission of key %s is updated as %s", r.Name, r.Perm.Key, authpb.Permission_Type_name[int32(r.Perm.PermType)])
+	return &pb.AuthRoleGrantPermissionResponse{}, nil
+}
+
+func (as *casbinAuthStore) isOpPermitted(userName string, revision uint64, key, rangeEnd []byte, permTyp authpb.Permission_Type) error {
+	if !as.s.isAuthEnabled() {
+		return nil
+	}
+
+	tx := as.s.be.BatchTx()
+	tx.Lock()
+	defer tx.Unlock()
+
+	user := getUser(tx, userName)
+	if user == nil {
+		plog.Errorf("invalid user name %s for permission checking", userName)
+		return ErrPermissionDenied
+	}
+
+	if as.enforcer.Enforce(userName, string(key), string(rangeEnd), authpb.Permission_Type_name[int32(permTyp)]) {
+		return nil
+	}
+
+	return ErrPermissionDenied
+}
+
+func (as *casbinAuthStore) IsPutPermitted(authInfo *AuthInfo, key []byte) error {
+	return as.isOpPermitted(authInfo.Username, authInfo.Revision, key, nil, authpb.WRITE)
+}
+
+func (as *casbinAuthStore) IsRangePermitted(authInfo *AuthInfo, key, rangeEnd []byte) error {
+	return as.isOpPermitted(authInfo.Username, authInfo.Revision, key, rangeEnd, authpb.READ)
+}
+
+func (as *casbinAuthStore) IsDeleteRangePermitted(authInfo *AuthInfo, key, rangeEnd []byte) error {
+	return as.isOpPermitted(authInfo.Username, authInfo.Revision, key, rangeEnd, authpb.WRITE)
+}
+
+func (as *casbinAuthStore) IsAdminPermitted(authInfo *AuthInfo) error {
+	if !as.s.isAuthEnabled() {
+		return nil
+	}
+	if authInfo == nil {
+		return ErrUserEmpty
+	}
+
+	tx := as.s.be.BatchTx()
+	tx.Lock()
+	defer tx.Unlock()
+
+	u := getUser(tx, authInfo.Username)
+	if u == nil {
+		return ErrUserNotFound
+	}
+
+	if !as.hasRootRole(u) {
+		return ErrPermissionDenied
+	}
+
+	return nil
+}
+
+func NewCasbinAuthStore(be backend.Backend, tp TokenProvider) *casbinAuthStore {
+	e := casbin.NewEnforcer("casbin_model.conf", "casbin_policy.csv")
```

Comment:
Is the intent that `casbin_policy.csv` holds user-configurable policies? If so, how is this configuration synchronized across all members? If one member rejects but another accepts then the cluster will be inconsistent,